src/platform/macos.rs

```diff
@@ -0,0 +1,178 @@
+
+    let mut last_position = PhysicalPosition::<f64>::new(0.0, 0.0);
+
+    events_loop.run(move |event, _, control_flow| {
+        *control_flow = ControlFlow::Poll;
+
+        match event {
+            Event::LoopDestroyed => {}
+            Event::WindowEvent { event, .. } => match event {
+                WindowEvent::Resized(size) => {
+                    metal_layer
+                        .set_drawable_size(CGSize::new(size.width as f64, size.height as f64));
+                    window.request_redraw()
+                }
+                WindowEvent::CloseRequested => *control_flow = ControlFlow::Exit,
+                WindowEvent::KeyboardInput {
+                    input:
+                        KeyboardInput {
+                            virtual_keycode: _,
+                            modifiers: _,
+                            ..
+                        },
+                    ..
+                } => {
+
+                }
+                WindowEvent::CursorMoved {
+                    position,
+                    ..
+                } => {
+                    last_position = position;
+                }
+                WindowEvent::MouseInput {
+                    state,
+                    button,
+                    modifiers: _,
+                    ..
+                } => {
+                    handle_click(last_position, state, button)
+                }
+                _ => (),
+            },
+            Event::RedrawRequested(_) => {}
+            _ => (),
+        }
+        if let Some(drawable) = metal_layer.next_drawable() {
+            let drawable_size = {
+                let size = metal_layer.drawable_size();
+                Size::new(size.width as scalar, size.height as scalar)
+            };
+            let mut surface = unsafe {
+                let texture_info = 
+                    mtl::TextureInfo::new(drawable.texture().as_ptr() as mtl::Handle);
+                let backend_render_target = BackendRenderTarget::new_metal(
+                    (drawable_size.width as i32, drawable_size.height as i32),
+                    1,
+                    &texture_info,
+                );
+                Surface::from_backend_render_target(
+                    &mut context,
+                    &backend_render_target,
+                    SurfaceOrigin::TopLeft,
+                    ColorType::BGRA8888,
+                    None,
+                    None,
+                )
+                .unwrap()
+    
+            };
+            handle_redraw(surface.canvas(), &mut tree);
+            surface.flush_and_submit();
+            drop(surface); 
+            let command_buffer = command_queue.new_command_buffer();
+            command_buffer.present_drawable(drawable);
+            command_buffer.commit();
+        }
+        
+    });
+}
+
+/// Renders a rectangle that occupies exactly half of the canvas
+#[cfg(all(target_os = "macos", feature = "metal"))]
```

Comment:
we should get rid of this draw function here. This is the job of the orchestrator.